common/numerated/src/numerated.rs

```diff
@@ -90,81 +68,134 @@ pub trait Bound<T: Sized>: From<T> + Copy {
 pub trait Numerated: Copy + Sized + Ord + Eq {
     /// Numerate type: type that describes the distances between two values of `Self`.
-    type N: PrimInt + Unsigned;
+    type Distance: PrimInt + Unsigned;
     /// Bound type: type for which any value can be mapped to `Self`,
     /// and also has __upper__ value, which is bigger than any value of `Self`.
-    type B: Bound<Self>;
+    type Bound: Bound<Self>;
     /// Adds `num` to `self`, if `self + num` is enclosed by `self` and `other`.
     ///
     /// # Guaranties
     /// - iff `self + num` is enclosed by `self` and `other`, then returns `Some(_)`.
     /// - iff `self.add_if_enclosed_by(num, other) == Some(a)`,
     /// then `a.sub_if_enclosed_by(num, self) == Some(self)`.
-    fn add_if_enclosed_by(self, num: Self::N, other: Self) -> Option<Self>;
+    fn add_if_enclosed_by(self, num: Self::Distance, other: Self) -> Option<Self>;
     /// Subtracts `num` from `self`, if `self - num` is enclosed by `self` and `other`.
     ///
     /// # Guaranties
     /// - iff `self - num` is enclosed by `self` and `other`, then returns `Some(_)`.
     /// - iff `self.sub_if_enclosed_by(num, other) == Some(a)`,
     /// then `a.add_if_enclosed_by(num, self) == Some(self)`.
-    fn sub_if_enclosed_by(self, num: Self::N, other: Self) -> Option<Self>;
-    /// Returns `self - other`, if `self ≥ other`.
+    fn sub_if_enclosed_by(self, num: Self::Distance, other: Self) -> Option<Self>;
+    /// Returns a distance between `self` and `other`
     ///
     /// # Guaranties
-    /// - iff `self ≥ other`, then returns `Some(_)`.
-    /// - iff `self == other`, then returns `Some(0)`.
-    /// - iff `self.distance(other) == Some(a)`, then
+    /// - iff `self == other`, then returns `0`.
+    /// - `self.distance(other) == other.distance(self)`.
+    /// - iff `self.distance(other) == a` and `self ≥ other` then
     ///   - `self.sub_if_enclosed_by(a, other) == Some(other)`
     ///   - `other.add_if_enclosed_by(a, self) == Some(self)`
-    fn distance(self, other: Self) -> Option<Self::N>;
+    fn distance(self, other: Self) -> Self::Distance;
     /// Increments `self`, if `self < other`.
     fn inc_if_lt(self, other: Self) -> Option<Self> {
-        self.add_if_enclosed_by(Self::N::one(), other)
+        self.add_if_enclosed_by(Self::Distance::one(), other)
     }
     /// Decrements `self`, if `self` > `other`.
     fn dec_if_gt(self, other: Self) -> Option<Self> {
-        self.sub_if_enclosed_by(Self::N::one(), other)
+        self.sub_if_enclosed_by(Self::Distance::one(), other)
     }
     /// Returns `true`, if `self` is enclosed by `a` and `b`.
     fn enclosed_by(self, a: &Self, b: &Self) -> bool {
         self <= *a.max(b) && self >= *a.min(b)
     }
 }
 
-impl<T> From<T> for BoundValue<T> {
+/// Bound type for `Option<T>`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, derive_more::From)]
+pub struct OptionBound<T>(Option<T>);
```

Comment:
Almost every of the following impls can be just derived

Reply:
proof of concept